File: clients/vscode/src/changelogService.ts
import { extractChangelogText, parseSections, selectRange } from './core/changelog';
import { ChangelogSection, Package } from './core/types';
import { PubDevApi } from './pub/pubDevApi';

/** Everything the changelog panel needs to render. */
export interface ChangelogRequest {
  packageName: string;
  fromVersion: string;
  toVersion: string;
  sections: ChangelogSection[];
  publishedAt: Map<string, Date>;
  /** Set when we could not match the changelog to the version range. */
  showingEverything: boolean;
}

/** A whole changelog can be hundreds of releases; nobody scrolls that far. */
const MAX_FALLBACK_SECTIONS = 20;

/**
 * Turns a package into something the changelog panel can render: fetch the
 * page, cut it into versions, keep the ones the update would bring in.
 *
 * Results are cached per package and version range, so reopening the same
 * changelog is instant while two projects pinning different versions of the
 * same package still each get their own.
 */
export class ChangelogService {
  private readonly cache = new Map<string, ChangelogRequest>();

  constructor(private readonly api: PubDevApi) {}

  async load(pkg: Package): Promise<ChangelogRequest> {
    const key = `${pkg.name}@${pkg.currentVersion}->${pkg.latestVersion}`;
    const cached = this.cache.get(key);
    if (cached) return cached;

    const request = await this.build(pkg);
    this.cache.set(key, request);
    return request;
  }

  clearCache(): void {
    this.cache.clear();
  }

  private async build(pkg: Package): Promise<ChangelogRequest> {
    const [html, remote] = await Promise.all([
      this.api.getChangelogHtml(pkg.name),
      this.api.getPackage(pkg.name)
    ]);

    const all = html ? parseSections(extractChangelogText(html)) : [];
    const inRange = selectRange(all, pkg.currentVersion, pkg.latestVersion);
    const showingEverything = inRange.length === 0 && all.length > 0;

    return {
      packageName: pkg.name,
      fromVersion: pkg.currentVersion,
      toVersion: pkg.latestVersion,
      sections: showingEverything ? all.slice(0, MAX_FALLBACK_SECTIONS) : inRange,
      publishedAt: remote?.publishedAt ?? new Map(),
      showingEverything
    };
  }
}

File: clients/intellij/src/main/kotlin/dev/pubgrade/ChangelogService.kt
package dev.pubgrade

import dev.pubgrade.core.ChangelogSection
import dev.pubgrade.core.Package
import dev.pubgrade.core.extractChangelogText
import dev.pubgrade.core.parseSections
import dev.pubgrade.core.selectRange
import dev.pubgrade.pub.PubDevApi
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

/** Everything the changelog panel needs to render. */
data class ChangelogRequest(
    val packageName: String,
    val fromVersion: String,
    val toVersion: String,
    val sections: List<ChangelogSection>,
    val publishedAt: Map<String, Instant>,
    /** Set when we could not match the changelog to the version range. */
    val showingEverything: Boolean
)

/** A whole changelog can be hundreds of releases; nobody scrolls that far. */
private const val MAX_FALLBACK_SECTIONS = 20

/**
 * Turns a package into something the changelog panel can render: fetch the
 * page, cut it into versions, keep the ones the update would bring in.
 *
 * Results are cached per package and version range, so reopening the same
 * changelog is instant while two projects pinning different versions of the
 * same package still each get their own.
 */
class ChangelogService(private val api: PubDevApi) {

    private val cache = ConcurrentHashMap<String, ChangelogRequest>()

    /** Blocking: call it from a background task, never the UI thread. */
    fun load(pkg: Package): ChangelogRequest {
        val key = "${pkg.name}@${pkg.currentVersion}->${pkg.latestVersion}"
        return cache.computeIfAbsent(key) { build(pkg) }
    }

    fun clearCache() {
        cache.clear()
    }

    private fun build(pkg: Package): ChangelogRequest {
        val html = api.getChangelogHtml(pkg.name)
        val remote = api.getPackage(pkg.name)

        val all = if (html != null) parseSections(extractChangelogText(html)) else emptyList()
        val inRange = selectRange(all, pkg.currentVersion, pkg.latestVersion)
        val showingEverything = inRange.isEmpty() && all.isNotEmpty()

        return ChangelogRequest(
            packageName = pkg.name,
            fromVersion = pkg.currentVersion,
            toVersion = pkg.latestVersion,
            sections = if (showingEverything) all.take(MAX_FALLBACK_SECTIONS) else inRange,
            publishedAt = remote?.publishedAt ?: emptyMap(),
            showingEverything = showingEverything
        )
    }
}
